File: test/verify_map.py
import os
import glob
import xml.etree.ElementTree as ET
import re

TMX_FILE = "test/maps.tmx"
BIN_PATTERN = "test/maps*_*.bin"

def verify():
    if not os.path.exists(TMX_FILE):
        print(f"TMX file not found: {TMX_FILE}")
        return

    print(f"Parsing {TMX_FILE}...")
    tree = ET.parse(TMX_FILE)
    root = tree.getroot()

    # Get properties
    properties = {}
    for prop in root.findall(".//property"):
        properties[prop.get("name")] = prop.get("value")

    if "screenWidth" not in properties or "screenHeight" not in properties:
        print("Error: screenWidth or screenHeight properties not found.")
        return

    screen_width = int(properties["screenWidth"])
    screen_height = int(properties["screenHeight"])
    print(f"Screen Size: {screen_width}x{screen_height}")

    # Parse chunks
    map_data = {} # (x, y) -> gid
    
    # We need to find the layer named 'map' or just the first tile layer
    # Looking at TMX content, layer id="1" name="map".
    layer = root.find(".//layer[@name='map']")
    if layer is None:
        layer = root.find(".//layer") # Fallback to first layer
    
    if layer is None:
        print("Error: No tile layer found.")
        return
        
    data = layer.find("data")
    if data is None:
        print("Error: No data in layer.")
        return
        
    if data.get("encoding") != "csv":
        print(f"Error: Unsupported encoding {data.get('encoding')}. Expected csv.")
        return

    chunks = data.findall("chunk")
    
    min_x, min_y, max_x, max_y = float('inf'), float('inf'), float('-inf'), float('-inf')

    for chunk in chunks:
        chunk_x = int(chunk.get("x"))
        chunk_y = int(chunk.get("y"))
        chunk_width = int(chunk.get("width"))
        chunk_height = int(chunk.get("height"))
        
        text = chunk.text.strip()
        gids = [int(x) for x in re.split(r'[,\s]+', text) if x]
        
        for i, gid in enumerate(gids):
            if gid == 0: continue
            
            # Mask flip bits
            gid &= 0x1FFFFFFF
            
            local_x = i % chunk_width
            local_y = i // chunk_width
            global_x = chunk_x + local_x
            global_y = chunk_y + local_y
            
            map_data[(global_x, global_y)] = gid
            
            min_x = min(min_x, global_x)
            min_y = min(min_y, global_y)
            max_x = max(max_x, global_x)
            max_y = max(max_y, global_y)

    print(f"Map Bounds: X=[{min_x}..{max_x}], Y=[{min_y}..{max_y}]")

    # Find binary files
    bin_files = glob.glob(BIN_PATTERN)
    if not bin_files:
        print(f"No binary files found matching {BIN_PATTERN}")
        # Try finding any .bin in test/ that looks like maps...
        return

    print(f"Found {len(bin_files)} binary files.")
    
    # Determine screensPerRow.
    # Exporter logic: maxX is calculated. 
    # Logic: screensPerRow = ceil((maxX + 1) / screenWidth).
    # Assuming map starts at 0,0 for export indexing.
    # If the TMX shows chunks starting at 0, then export index corresponds to global coords.
    
    # Check if we have data at negative coordinates.
    if min_x < 0 or min_y < 0:
        print("Warning: Map has data at negative coordinates. Exporter logic assumes 0-based?")
    
    # Use max_x from parsing to determine geometry
    screens_per_row = (max_x // screen_width) + 1 # Equivalent to ceil((max_x+1)/screen_width) if max_x is 0-indexed index.
    # Actually: ceil((max_x + 1) / width)
    # If max_x = 31, width=32 -> 32/32 = 1.
    # If max_x = 32, width=32 -> 33/32 = 2.
    import math
    screens_per_row = math.ceil((max_x + 1) / screen_width)
    print(f"Calculated screensPerRow: {screens_per_row}")

    all_passed = True

    for bin_file in bin_files:
        # Extract index
        # Expecting e.g. maps_0.bin or maps._0.bin
        # Regex to find last number before .bin
        match = re.search(r'_(\d+)\.bin$', bin_file)
        if not match:
            print(f"Skipping {bin_file}: Cannot parse screen index.")
            continue
            
        screen_index = int(match.group(1))
        
        with open(bin_file, "rb") as f:
            content = f.read()
            
        if len(content) != screen_width * screen_height:
            print(f"[FAIL] {bin_file}: Size {len(content)} != Expected {screen_width*screen_height}")
            all_passed = False
            continue
            
        # Verify content
        screen_y = screen_index // screens_per_row
        screen_x = screen_index % screens_per_row
        
        start_x = screen_x * screen_width
        start_y = screen_y * screen_height
        
        file_passed = True
        for i, byte_val in enumerate(content):
            lx = i % screen_width
            ly = i // screen_width
            
            gx = start_x + lx
            gy = start_y + ly
            
            expected_gid = map_data.get((gx, gy), 0)
            
            # TMX uses GIDs (1-based usually). Exporter writes TileIDs (0-based).
            # We assume firstgid=1.
            # GID 0 (empty) -> 0 in binary (default Uint8)
            # GID 1 (ID 0)  -> 0 in binary
            # GID X         -> X - 1 in binary
            
            if expected_gid == 0:
                expected_val = 0
            else:
                expected_val = expected_gid - 1
            
            if byte_val != expected_val:
                print(f"[FAIL] {bin_file} at ({lx},{ly}): Found {byte_val}, Expected {expected_val} (GID {expected_gid})")
                file_passed = False
                break
        
        if file_passed:
            print(f"[PASS] {bin_file} verified.")
        else:
            all_passed = False

    if all_passed:
        print("\nSUCCESS: All files verified against TMX.")
    else:
        print("\nFAILURE: Verification failed.")

if __name__ == "__main__":
    verify()

File: test/verify_map.js
const fs = require('fs');
const path = require('path');

// Configuration
const TMX_FILE = path.join(__dirname, 'maps.tmx');
const BIN_BASE_NAME = path.join(__dirname, 'maps'); // Expects maps_0.bin, maps_1.bin...

function parseTMX(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');

    // Extract screen dimensions
    const screenWidthMatch = content.match(/<property name="screenWidth" type="int" value="(\d+)"\/>/);
    const screenHeightMatch = content.match(/<property name="screenHeight" type="int" value="(\d+)"\/>/);

    if (!screenWidthMatch || !screenHeightMatch) {
        throw new Error("Could not find screenWidth or screenHeight properties in TMX.");
    }

    const screenWidth = parseInt(screenWidthMatch[1]);
    const screenHeight = parseInt(screenHeightMatch[1]);

    console.log(`Expected Screen Size: ${screenWidth}x${screenHeight}`);

    // Parse Chunks
    // Regex to find chunks
    const chunkRegex = /<chunk x="(-?\d+)" y="(-?\d+)" width="(\d+)" height="(\d+)">([\s\S]*?)<\/chunk>/g;
    let match;
    const mapData = {}; // key: "x,y" (global), value: gid

    let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;

    while ((match = chunkRegex.exec(content)) !== null) {
        const chunkX = parseInt(match[1]);
        const chunkY = parseInt(match[2]);
        const chunkW = parseInt(match[3]);
        const chunkH = parseInt(match[4]);
        const chunkContent = match[5].trim();

        const gids = chunkContent.split(/[,\s]+/).map(s => parseInt(s)).filter(n => !isNaN(n));

        for (let i = 0; i < gids.length; i++) {
            const localX = i % chunkW;
            const localY = Math.floor(i / chunkW);
            const globalX = chunkX + localX;
            const globalY = chunkY + localY;

            // Mask out flip bits (high 3 bits)
            // 0xFFFFFFFF >>> 0 to ensure unsigned in JS logic if needed, but here simple mask works
            const gid = gids[i] & 0x1FFFFFFF;

            if (gid !== 0) { // Store only non-empty
                mapData[`${globalX},${globalY}`] = gid;

                if (globalX > maxX) maxX = globalX;
                if (globalY > maxY) maxY = globalY;
                if (globalX < minX) minX = globalX;
                if (globalY < minY) minY = globalY;
            }
        }
    }

    console.log(`Map Bounds (Chunks): X=[${minX}..${maxX}], Y=[${minY}..${maxY}]`);
    return { screenWidth, screenHeight, mapData, maxX, maxY };
}

function verify() {
    console.log("Starting verification...");

    if (!fs.existsSync(TMX_FILE)) {
        console.error("TMX file not found:", TMX_FILE);
        return;
    }

    const { screenWidth, screenHeight, mapData, maxX, maxY } = parseTMX(TMX_FILE);

    // Determine expected number of screens
    // The export script logic:
    // while (cellAt(maxX, maxY) != -1)... starts at 0,0?
    // tmx bounds might not start at 0.
    // Assuming map starts at 0,0 for export logic (since it iterates x=0, y=0...)

    // We'll trust the exporter's file naming: maps_0.bin, maps_1.bin...
    // We check locally for files matching pattern.

    const dir = path.dirname(BIN_BASE_NAME);
    const base = path.basename(BIN_BASE_NAME);
    const files = fs.readdirSync(dir).filter(f => f.startsWith(base + '_') && f.endsWith('.bin'));

    if (files.length === 0) {
        console.log("No exported binary files found. Please export 'maps.tmx' to 'maps.bin' (generating maps_*.bin) first.");
        return;
    }

    console.log(`Found ${files.length} binary files to verify.`);

    // Re-calculate screensPerRow to decode screen index
    // We assume export max is same as our parsed max, BUT exporter calculates it by scanning.
    // If we assume the map goes from 0 to maxX found in chunks.
    const effectiveMaxX = Math.max(0, maxX); // Ensure non-negative? 
    // Actually, if chunks start at 0, maxX is accurate.

    const screensPerRow = Math.ceil((effectiveMaxX + 1) / screenWidth);
    console.log(`Calculated screensPerRow: ${screensPerRow}`);

    let allPassed = true;

    files.forEach(file => {
        const filePath = path.join(dir, file);
        const match = file.match(/_(\d+)\.bin$/);
        if (!match) return;

        const screenIndex = parseInt(match[1]);
        const content = fs.readFileSync(filePath);

        if (content.length !== screenWidth * screenHeight) {
            console.error(`[FAIL] ${file}: File size ${content.length} does not match expected ${screenWidth * screenHeight}`);
            allPassed = false;
            return;
        }

        // Determine screen position
        const screenY = Math.floor(screenIndex / screensPerRow);
        const screenX = screenIndex % screensPerRow;

        const startX = screenX * screenWidth;
        const startY = screenY * screenHeight;

        // Verify content
        for (let i = 0; i < content.length; i++) {
            const val = content[i];

            const lx = i % screenWidth;
            const ly = Math.floor(i / screenWidth);

            const gx = startX + lx;
            const gy = startY + ly;

            const expectedGid = mapData[`${gx},${gy}`] || 0;

            // Note: Expected GID 0 means empty. 
            // Binary might have 0 for empty?
            // Also need to handle tile ID mapping if the exporter uses local ID.
            // But we assume GID match for now.

            // NOTE: Tiled's tileId -1 (if that's what it returns for empty) vs 0.
            // If the script writes tileId, and tileId is GID...
            // Wait, if tileId is -1 for empty, writing it to Uint8 might wrap to 255?
            // The script: `if (tile.tileId !== -1) screens[...]=tileId`
            // Else `screens` is initialized to 0 (default Uint8Array).
            // So empty is 0.
            // So if mapData has 0, value should be 0.

            if (val !== expectedGid) {
                // Heuristic for local ID vs GID mismatch
                if (val === expectedGid - 1) {
                    // Maybe it exported local ID?
                    console.error(`[FAIL] ${file} at local(${lx},${ly}) global(${gx},${gy}): Value ${val} != Expected ${expectedGid}. (Looks like GID-1?)`);
                    allPassed = false;
                    return; // Report first error per file to avoid spam
                }

                console.error(`[FAIL] ${file} at local(${lx},${ly}) global(${gx},${gy}): Value ${val} != Expected ${expectedGid}`);
                allPassed = false;
                // return; 
            }
        }
        console.log(`[PASS] ${file} verified.`);
    });

    if (allPassed) {
        console.log("TYPE: SUCCESS. All binary files match TMX data.");
    } else {
        console.log("TYPE: FAILURE. Some files did not match.");
    }
}

verify();
